arch/x86/entry.S

```diff
@@ -145,10 +145,9 @@ ENTRY(enter_usermode)
     PUSHF
 
     /* Save stack pointer onto per-cpu */
-    mov %_ASM_SP, %gs:usermode_private
+    mov %_ASM_DX, %gs:(usermode_private)
 
-    /* Move to user stack */
-    mov %_ASM_DX, %_ASM_SP
+    syscall_to_usermode
```

Comment:
`syscall_to_usermode` also switches GS and, more importantly, CR3 to `user_cr3`. I guess that's an intended (and orthogonal) bugfix?

Reply:
Yes (to some extent). The previous version was only changing CR3 and swapgs but after the switch to user stack modification. Which required that user stack was also mapped into kernel page table. With this change we fix two things:

1. Lift the above requirement 
2. Store user stack into percpu and properly switch between kernel and user stack later on